xla/pjrt/c/pjrt_c_api_gpu_internal.cc

```diff
@@ -83,6 +83,8 @@ PJRT_Error* PJRT_Client_Create(PJRT_Client_Create_Args* args) {
           {"node_id", PJRT_NamedValue_Type::PJRT_NamedValue_kInt64},
           {"num_nodes", PJRT_NamedValue_Type::PJRT_NamedValue_kInt64},
           {"enable_mock_nccl", PJRT_NamedValue_Type::PJRT_NamedValue_kBool},
+          {"mock_num_hosts_per_slice",
```

Comment:
Probably makes more sense to have a single flag, with expected GPU topology? I'm also not sure what "slice" means, if we are talking about a number of hosts per slice?

Reply:
Slice means a node/machine here.

For instance, if you have 8 nodes, each with 8 GPUs and those 8 GPUs are split between 2 processes on each node, then

number of slices = 8
number of hosts per slice = 2
number of devices per host = 4

The naming is taken from here: https://github.com/openxla/xla/blob/9544c5184db44e05b77b46637c3c98af812f26f5/xla/pjrt/gpu/gpu_topology.proto#L25

How do you imagine the single flag to look like? Something like mock_gpu_topology=8x2? Or do you want to encode the entire GPU topology proto in the flag? I.e. something like mock_gpu_topology=8x2x4x80xNVIDIA A100-SXM4-40GB (not a big fan of this one).

Reply:
> Slice means a node/machine here.

I'm not sure it does; @hawkinsp WDYT? Since we've discussed these namings recently.

> Something like mock_gpu_topology=8x2?

Exactly. Also given that hosts are organized into slices potentially up to 8x8x8

> I.e. something like mock_gpu_topology=8x2x4x80xNVIDIA A100-SXM4-40GB (not a big fan of this one).

This needs to be a separate flag, and I think orthogonal to this change.

I think it makes sense to have two modes:

 - Fully device-less mocking: does not exist now, will require supplying device spec with a flag
 - Mocking from 0th GPU: just specify topology.

Reply:
Currently, devices will be assigned the same slice index iff [their boot_id is the same](https://github.com/openxla/xla/blob/6fabf593785ae1391c40d01d60618c249ce5c9ea/xla/pjrt/distributed/topology_util.cc#L140). So in essence, a slice is currently implemented to correspond to the machine. I do realize that this might not be consistent with the [comment on the proto](https://github.com/openxla/xla/blob/6fabf593785ae1391c40d01d60618c249ce5c9ea/xla/pjrt/gpu/gpu_topology.proto#L23) "Devices on the same slice are connected by the fast network via NVLinks, which could be within a host or span across multiple hosts".

As for the `mock_gpu_topology`, should this replace the `num_nodes` parameter? Or are you imagining this would supersede `num_nodes`? (I.e., if topology is specified, then num-nodes is ignored or at least force to be consistent with the topology.) Currently, the code for building GPU topology assumes that [nodes and hosts are the same thing](https://github.com/openxla/xla/blob/6fabf593785ae1391c40d01d60618c249ce5c9ea/xla/pjrt/distributed/topology_util.cc#L259). Or are we assuming that hosts are not necessarily the same thing as nodes, so we just stamp out the GPU topology without any checks (and also bypassing the global topology proto)?

I am currently leaning towards creating the GPU topology from the provided mock topology and not insisting on num_nodes==num_hosts_per_slice*num_slices (even though you will then compile with topology that will never exist in practice, so the cache will never hit).

Reply:
> So in essence, a slice is currently implemented to correspond to the machine

As you specify further down, that seems to contradict the meaning of the word "slice" as used already in PjRT.

At least intuitively, I would interpret `--mock-gpu-topology=8x8` as 8 hosts with 8 GPUs each (and 8 slices I'd guess?)

> As for the mock_gpu_topology, should this replace the num_nodes parameter

Ideally the proto need to be redone, but I think the first number from mock GPU topology should specify num_nodes.

I'd ignore num_slices entirely, or just assume it's always equal to the number of hosts for the initial implementation.